File: model_utils.py
# model_utils.py

import numpy as np
import math
import 합성곱
from 오차역전파법 import Variable

def load_model(path="model.npy"):
    """
    np.save로 저장된 dict를 load하여 반환합니다.
    {
      "conv_kernels": list of 8×3×3 floats,
      "conv_biases":  list of 8 floats,
      "fc_w":         list of feat_dim floats,
      "fc_b":         float
    }
    """
    data = np.load(path, allow_pickle=True).item()
    return data

def predict(image: np.ndarray, model_path="model.npy") -> float:
    """
    단일 9×9 바이너리 이미지에 대해 학습된 모델을 실행하고
    0~1 사이의 확률(prediction)을 반환합니다.
    
    image: numpy.ndarray, shape (9,9), 값 0 or 1
    model_path: 저장된 model.npy 경로
    """
    # 1) 모델 불러오기
    mdl = load_model(model_path)
    conv_kernels = mdl["conv_kernels"]   # shape [C,3,3]
    conv_biases  = mdl["conv_biases"]    # length C
    fc_w         = mdl["fc_w"]           # length feat_dim
    fc_b         = mdl["fc_b"]           # scalar

    C = len(conv_kernels)
    
    # 2) Variable 객체로 래핑
    img_var = [[Variable(float(p)) for p in row] for row in image]
    kern_vars = [
        [[Variable(conv_kernels[c][i][j]) for j in range(3)] for i in range(3)]
        for c in range(C)
    ]
    bias_vars = [Variable(b) for b in conv_biases]
    fc_w_vars  = [Variable(w) for w in fc_w]
    fc_b_var   = Variable(fc_b)

    # 3) Conv → ReLU → MaxPool → Flatten
    feats = []
    for kern, b in zip(kern_vars, bias_vars):
        fm = 합성곱.convolution(img_var, kern, b)          # 7×7
        fm = [[v.relu() for v in row] for row in fm]       # ReLU
        pm = 합성곱.max_pooling(fm, pool_size=2)           # 3×3
        for row in pm:
            feats.extend(row)                              # flatten

    # 4) FC → Sigmoid
    logit = Variable(0.0)
    for w_var, x_var in zip(fc_w_vars, feats):
        logit = logit + w_var * x_var
    logit = logit + fc_b_var
    pred = logit.sigmoid()

    return pred.data

File: web/model.utils.js
// model_utils.js

async function loadModel(path = "model.json") {
  const resp = await fetch(path);
  const mdl = await resp.json();
  console.log("🔍 loaded model keys:", Object.keys(mdl));
  console.log(
    "🔍 conv_kernels shape:",
    mdl.conv_kernels.length,
    mdl.conv_kernels[0].length,
    mdl.conv_kernels[0][0].length
  );
  return {
    convKernels: mdl.conv_kernels, // [C][3][3]
    convBiases: mdl.conv_biases, // [C]
    fcW: mdl.fc_w, // [featDim]
    fcB: mdl.fc_b, // scalar
  };
}

function relu(x) {
  return x > 0 ? x : 0;
}
function sigmoid(x) {
  return 1 / (1 + Math.exp(-x));
}

function conv2d(image, kernel, bias) {
  const H = image.length,
    W = image[0].length,
    K = kernel.length;
  const out = [];
  for (let y = 0; y <= H - K; y++) {
    const row = [];
    for (let x = 0; x <= W - K; x++) {
      let sum = 0;
      for (let i = 0; i < K; i++)
        for (let j = 0; j < K; j++) sum += image[y + i][x + j] * kernel[i][j];
      row.push(sum - bias);
    }
    out.push(row);
  }
  return out;
}

function maxPool2d(matrix, size = 2, stride = 2) {
  const H = matrix.length,
    W = matrix[0].length;
  const oh = Math.floor((H - size) / stride) + 1;
  const ow = Math.floor((W - size) / stride) + 1;
  const pooled = Array.from({ length: oh }, () => Array(ow).fill(0));
  for (let i = 0; i < oh; i++) {
    for (let j = 0; j < ow; j++) {
      let m = -Infinity;
      for (let di = 0; di < size; di++)
        for (let dj = 0; dj < size; dj++)
          m = Math.max(m, matrix[i * stride + di][j * stride + dj]);
      pooled[i][j] = m;
    }
  }
  return pooled;
}

async function predict(image, modelPath = "model.json") {
  // image: 9×9 array of 0/1
  const { convKernels, convBiases, fcW, fcB } = await loadModel(modelPath);

  console.log("🔍 First kernel:", convKernels[0]); // kernel[0] 실제 값 체크

  const C = convKernels.length;
  let feats = [];

  for (let c = 0; c < C; c++) {
    // 1) conv
    let fm = conv2d(image, convKernels[c], convBiases[c]);
    console.log(`🖼︎ feature map ${c}[0][0]=`, fm[0][0]);
    // 2) relu
    fm = fm.map((row) => row.map(relu));
    // 3) pool
    const pm = maxPool2d(fm, 2, 2);
    console.log(`🔢 pooled ${c}[0][0]=`, pm[0][0]);
    // 4) flatten
    for (const row of pm) feats.push(...row);
  }

  // 5) FC → logit
  let logit = feats.reduce((acc, v, i) => acc + v * fcW[i], 0) + fcB;
  console.log("📊 logit:", logit);
  const prob = sigmoid(logit);
  console.log("🎯 sigmoid(logit):", prob);

  return prob;
}
